Document array slices. (#7911)
ethereum/solidity: docs/types/reference-types.rst

        }
    }

.. index:: ! array;slice

.. _array-slices:

Array Slices
------------

> Array slices are a view on a contigous portion of an array.

Reviewer: ```suggestion
Array slices are a view on a contiguous portion of an array.
```

They are written as ``x[start:end]``, where ``start`` and
``end`` are expressions resulting in a uint256 type (or
implicitly convertible to it). The first element of the
slice is ``x[start]`` and the last element is ``x[end - 1]``.

If ``start`` is greater than ``end`` or if ``end`` is greater
than the length of the array, an exception is thrown.

> Both ``start`` and ``end`` are optional, and they default to

Reviewer: ```suggestion
Both ``start`` and ``end`` are optional: ``start`` defaults to ``0`` and ``end`` defaults to the length of the array.
```
?
But also fine as it is.

``0`` and the length of the array, respectively.

> Array slices do not have any members. They are implicitly

Reviewer: Well the "implicitly convertible to arrays of their underlying type" is true for calldata arrays - but I'm not yet sure what will happen for memory or storage... but we can adjust this once that's being worked on... maybe we'll have ``copyof`` by then, then that will be needed for the conversion...

Reviewer 2: Okay, if I understood correctly, we leave this as it is for now, changing it as required when an implementation for memory/storage array slices is available.

convertible to arrays of their underlying type

> and support index access. Index access is of course

Reviewer 3: ```suggestion
and support index access. Index access is
```

Not really needed

not absolute in the underlying array, but relative to
the start of the slice.

Array slices do not have a type name which means
no variable can have an array slices as type,
they only exist in intermediate expressions.

.. note::
    As of now, array slices are only implemented for calldata arrays.

Array slices are useful to ABI-decode secondary data passed in function parameters:

::

    pragma solidity >=0.4.99 <0.7.0;

    contract C {

>         function verifyBet(address recipient, bytes calldata _data) {

Reviewer: ```suggestion
        function verifyBet(address recipient, bytes calldata _data) external {
```

>             bytes4 sig = _data[:4];

Reviewer: ```suggestion
            bytes4 sig = abi.decode(_data[:4], (bytes4));
```
Otherwise it won't be convertible...

Reviewer: Since there are no ``calldata`` variables yet, that's practically the only use case for them. The only other thing is *explicit* conversion to ``memory`` types:
``bytes memory tail = bytes(_data[4:]);``

.. _array-slices:

Array Slices
------------

Array slices are a view on a contiguous portion of an array.
They are written as ``x[start:end]``, where ``start`` and
``end`` are expressions resulting in a uint256 type (or
implicitly convertible to it). The first element of the
slice is ``x[start]`` and the last element is ``x[end - 1]``.

If ``start`` is greater than ``end`` or if ``end`` is greater
than the length of the array, an exception is thrown.

Both ``start`` and ``end`` are optional: ``start`` defaults
 to ``0`` and ``end`` defaults to the length of the array.

Array slices do not have any members. They are implicitly
convertible to arrays of their underlying type
and support index access. Index access is not absolute
in the underlying array, but relative to the start of
the slice.

Array slices do not have a type name which means
no variable can have an array slices as type,
they only exist in intermediate expressions.

.. note::
    As of now, array slices are only implemented for calldata arrays.

Array slices are useful to ABI-decode secondary data passed in function parameters:

::

    pragma solidity >=0.4.99 <0.7.0;

    contract C {
        address owner;

>         modifier atleastPayloadSize(uint size) {

Author: abi.decode actually does that check.

Reviewer 2: Removed it

            require(msg.data.length >= size + 4);
            _;
        }

        modifier onlySelf() {
            require(msg.sender == address(this));
            _;
        }

        function setOwner(address _owner) onlySelf public {
            owner = _owner;
        }

        function forward(bytes calldata _payload) atleastPayloadSize(20) external {

>             bytes4 sig = abi.decode(_payload[:4], (bytes4));

Author: Is this the only way to convert?

Reviewer 2: If I do

```
bytes4 sig = bytes4(_payload[:4]);
```

I get the following error

```
Error: Explicit type conversion not allowed from "bytes calldata slice" to "bytes4".
...
```

            if (sig == bytes4(keccak256("setOwner(address)"))) {
                address ownerAddr = abi.decode(_payload[4:], (address));
                require(ownerAddr != address(0), "Cannot set owner to zero.");
            }

>             (bool status,) = address(this).delegatecall(_payload);

Author: Can you forward it to some other address stored here in storage?

Author: And maybe just use a regular call.

Author: Ah sorry, did not see "onlySelf".

Hm, still, this looks like a protection that could otherwise be done easily on 'setOwner' itself.

Maybe a better example would be to have a secondary contract put in front of another and stored there as manager/owner (but that is the part of the example we do not implement).

Reviewer 2: How about this example?

```
contract Proxy {
	// Address of the client contract managed by proxy i.e., this contract
	address client;

	constructor(address _client) public {
		client = _client;
	}
	
	// Forward call to "setOwner(address)" that is implemented by client
	// after doing basic validation on the address argument.
	function forward(bytes calldata _payload) external {
		bytes4 = abi.decode(_payload[:4], (bytes4));
		if (sig == bytes4(keccak256("setOwner(address)"))) {
			address owner = abi.decode(_payload[4:], (address));
			require(owner != address(0), "Address of owner cannot be zero.");
		}
		(bool status,) = client.delegatecall(_payload);
		require(status, "Forwarded call failed.");
	}
}
```

Author: Yep, looks good! Can you change the double slashes to triple slashes?

Reviewer 2: Updated PR. External tests are still failing though
